### PR TITLE
Added function util_extract_2d, which extracts spectra while keeping spatial information

### PR DESCRIPTION
**Pull request by ricolandman**

### CODE REVIEW

File: pycrires/pipeline.py [hunk 1]
```diff
@@ -3884,6 +3884,140 @@ def correct_wavelengths(
...
+
+        self._print_section("Process nodding frames", recipe_name="cr2res_util_extract_2d")
+
+        fits_files = sorted(list(self.file_dict[f'OBS_NODDING_COMBINED{nod_ab}'].keys()))
+
+        # Iterate over nod A exposures
+        for count_exp,fits_file in enumerate(fits_files):
+            
+            # Prepare SOF file
+            sof_file = pathlib.Path(self.product_folder / f"files_{count_exp:03d}.sof")
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3930]: I have changed the output folder of `obs_nodding` to "product/obs_nodding". I think that we can do the same for `util_extract_2d`, so change the output folder to "product/util_extract_2d".

File: pycrires/pipeline.py [hunk 2]
```diff
@@ -3884,6 +3884,140 @@ def correct_wavelengths(
...
+            sof_open = open(sof_file, "w", encoding="utf-8")
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3931]: Could be changed to a context manager (i.e. using ``with``)

File: pycrires/pipeline.py [hunk 3]
```diff
@@ -3884,6 +3884,140 @@ def correct_wavelengths(
...
+            sof_open.write(f"{fits_file} OBS_NODDING_OTHER\n")
+            tw_file = self.product_folder / f"cr2res_obs_nodding_trace_wave_{nod_ab}_{count_exp:03d}.fits"
+            sof_open.write(f"{tw_file} UTIL_WAVE_TW\n")
+            sof_open.close()
+
+            # Get data on size of the slit on the detector
+            tw_data = fits.open(tw_file)[1].data
+            slit_range = tw_data['SlitFraction'][0]
+            num_orders = tw_data['All'].shape[0]
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3940]: This works, but maybe clearer as `num_orders = tw_data['Order'].size`

File: pycrires/pipeline.py [hunk 4]
```diff
@@ -3884,6 +3884,140 @@ def correct_wavelengths(
...
+            tot_slit_fraction = slit_range[-1] - slit_range[0]
+            full_slit_length = 10 #arcsec
+            
+            # If extraction_length not specified, calculate the pixel scale
+            if extraction_length == None:
+                slit_fraction_pixels = np.mean(tw_data['Upper']-tw_data['Lower'])
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3946]: Add a comment what the 'Upper' and 'Lower' columns are? I wasn't sure. Also, the standard deviation of this difference seems rather large so I wonder how accurate the calculated pixel scale will be. Can we not adopt the actual pixel scale from the manual?

File: pycrires/pipeline.py [hunk 5]
```diff
@@ -3884,6 +3884,140 @@ def correct_wavelengths(
...
+                pixel_scale = full_slit_length * tot_slit_fraction  / slit_fraction_pixels
+                extraction_length = pixel_scale
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3948]: I think that this line can be removed when changing `pixel_scale` to `extraction_length` in the previous line

File: pycrires/pipeline.py [hunk 6]
```diff
@@ -3884,6 +3884,140 @@ def correct_wavelengths(
...
+
+            # Calculate the spatial size for the extraction and
+            # the number of spatial points
+            extraction_fraction = extraction_length / full_slit_length
+            N_points = spatial_oversampling * tot_slit_fraction / extraction_fraction
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3953]: It seems that `N_points` is the same as `spatial_oversampling` times `slit_fraction_pixels` (but please check), so this line can be simplified I think

File: pycrires/pipeline.py [hunk 7]
```diff
@@ -3884,6 +3884,140 @@ def correct_wavelengths(
...
+
+            # Round N_points up to the next odd integer
+            N_points = int(np.ceil(N_points) // 2 * 2 + 1)
+            
+
+            # Create EsoRex configuration file if not found
+            self._create_config("cr2res_util_extract", "util_extract", verbose)
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3960]: This can be moved before the `for` loop I think

File: pycrires/pipeline.py [hunk 8]
```diff
@@ -3884,6 +3884,140 @@ def correct_wavelengths(
...
+                      f'{lower_lim:.3f} - {upper_lim:.3f}')
+                
+                # Run EsoRex
+                esorex = [
+                    "esorex",
+                    f"--recipe-config={config_file}",
+                    f"--output-dir={self.product_folder}",
+                    "cr2res_util_extract",
+                    f"--slit_frac={lower_lim},{upper_lim}",
+                    f"--oversample={util_extract_oversample}",
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3989]: I would prefer not to overwrite esorex parameters from the command line, unless needed. I would therefore remove the ``util_extract_oversample`` parameter from this method and instead overwrite the default value with 12 in the configuration file. Then, if the user wants to change the value, that will be possible by adjusting the configuration file. Have a look at `_create_config` with the `obs_nodding` recipe where I had also change the oversampling to 12.

File: pycrires/pipeline.py [hunk 9]
```diff
@@ -3884,6 +3884,140 @@ def correct_wavelengths(
...
+
+                subprocess.run(esorex, cwd=self.product_folder, stdout=stdout, check=True)
+
+                if not verbose:
+                    print(" [DONE]\n")
+                
+                # Load results
+                hdu = fits.open(self.product_folder / f"cr2res_obs_nodding_combined{nod_ab}_{count_exp:03d}_extr1D.fits")
+                for det in np.arange(1, 4):
+                    for order in np.arange(2,9):
```
Comment (tomasstolker) [inline pycrires/pipeline.py:4007]: Here you could run the `for` loop over `tw_data['Order']` I think, which may then also work for other wavelengths settings where the number of orders might be different

**Review by tomasstolker — CHANGES_REQUESTED** [review]
Thanks a lot @ricolandman ! Again very nice work! My extraction is still running but seems to run smoothly so far 👍. I have added some comments for further improvements. Could you also pull the remote `main` to sync with your PR branch? Because I have changed the output location of the `obs_nodding`.

File: pycrires/pipeline.py (hunk 1 above)
Comment (ricolandman) [reply pycrires/pipeline.py]: Done

File: pycrires/pipeline.py (hunk 2 above)
Comment (ricolandman) [reply pycrires/pipeline.py]: Done

File: pycrires/pipeline.py (hunk 3 above)
Comment (ricolandman) [reply pycrires/pipeline.py:3982]: I am not sure that works, as the .size of a 2d array is different than the size of its first axis.

File: pycrires/pipeline.py (hunk 4 above)
Comment (ricolandman) [reply pycrires/pipeline.py]: Removed this part, as I now use the pixel scale from the manual (59 mas/pixel)

File: pycrires/pipeline.py (hunk 5 above)
Comment (ricolandman) [reply pycrires/pipeline.py]: Removed this part, as I now use the pixel scale from the manual (59 mas/pixel)

File: pycrires/pipeline.py (hunk 6 above)
Comment (ricolandman) [reply pycrires/pipeline.py:3989]: With the above change, I no longer calculate slit_fraction_pixels

File: pycrires/pipeline.py (hunk 7 above)
Comment (ricolandman) [reply pycrires/pipeline.py]: Done

File: pycrires/pipeline.py (hunk 8 above)
Comment (ricolandman) [reply pycrires/pipeline.py]: I think I have done this now, but please check if this is indeed the way you want it

File: pycrires/pipeline.py (hunk 9 above)
Comment (ricolandman) [reply pycrires/pipeline.py]: Done

File: pycrires/pipeline.py (hunk 3 above)
Comment (tomasstolker) [reply pycrires/pipeline.py:3982]: I think that the shape `tw_data['Order']` will be 1D but fine to leave it like this!

File: pycrires/pipeline.py (hunk 8 above)
Comment (tomasstolker) [reply pycrires/pipeline.py]: Thanks, yes that looks good!

**Review by tomasstolker — APPROVED** [review]

File: pycrires/pipeline.py (hunk 3 above)
Comment (ricolandman) [reply pycrires/pipeline.py:3982]: You're right, my bad, was thinking about the shape of tw_data['All'].

**Pull request merged**